certora/specs/YieldStreams.spec

```diff
@@ -0,0 +1,806 @@
+    require vault.balanceOf(alice) >= _aliceShares;
+
+    require to_mathint(bobsPrincipal) == (multiple + desloc) * (10 ^ 18);
+    uint256 _bobShares = vault.convertToShares(bobsPrincipal);
+    uint256 bobStreamId = aliceStreamId;
+    require streamIdToReceiver(bobStreamId) == carol;
+    require to_mathint(receiverTotalPrincipal(carol)) >= alicesPrincipal + bobsPrincipal;
+    require receiverPrincipal(carol, bobStreamId) == bobsPrincipal;
+    require vault.balanceOf(bob) >= _bobShares;
+
+
+    // add X% profit to vault
+    require 1 <= desloc;
+    require to_mathint(receiverTotalShares(carol)) >= _aliceShares + _bobShares + (multiple + desloc) * (10 ^ 18);
+    require to_mathint(vault.balanceOf(currentContract)) >= _aliceShares + _bobShares + (multiple + desloc) * (10 ^ 18);
+    
+    assert delta(to_mathint(previewClaimYield(carol)), alicesPrincipal + bobsPrincipal) <= 1;
+
+    uint256 claimed = claimYield(e, carol);
+
+    assert to_mathint(claimed) == alicesPrincipal + bobsPrincipal &&
+     asset.balanceOf(carol) == claimed &&
+     previewClaimYield(carol) == 0;
+}
+
+
+/**
+ * @Rule Verify property for the `claimYieldInShares` function to self
+ * @Category High
+ * @Description Ensures that the `claimYieldInShares` function verifies the correct behavior of the claimYieldInShares function when the receiver (bob) claims the yield in the form of shares to their own address. It sets up a scenario where a yield stream is opened between alice (the streamer) and bob (the receiver), with a specified principal amount. The rule then generates yield by adding a percentage of the principal to the vault's balance. Finally, it invokes the claimYieldInShares function with bob as the receiver and asserts that the claimed yield in shares is correctly calculated and transferred to bob's share balance, while ensuring that the contract's state is updated accordingly.
+ */
+rule verify_claimYieldInShares_toSelf(address alice, address bob, uint256 _principal, uint8 parts) {
+    env e;
+    uint256 _shares = vault.convertToShares(_principal);
+    require alice != 0 && bob != 0 && alice != bob && bob == e.msg.sender;
+
+    // State changes
+    uint256 streamId = nextStreamId();
+    require streamIdToReceiver(streamId) == bob;
+    require receiverTotalShares(bob) == _shares;
+    require receiverTotalPrincipal(bob) == _principal;
+    require receiverPrincipal(bob, streamId) == _principal;
+    require _shares > 0;
+    require vault.balanceOf(alice) >= _shares;
+    require 1 <= parts;
+    uint256 principalParts;
+    require to_mathint(principalParts) == _principal / parts;
+    uint256 expectedYieldInShares = vault.convertToShares(principalParts);
+    // Add X% profit to vault
+    require to_mathint(vault.balanceOf(currentContract)) == _shares + principalParts;
+
+    uint256 claimed = claimYieldInShares(e, bob);
+
+    assert delta(claimed, expectedYieldInShares) <= 1 &&
```

Comment:
perhaps multiple asserts is easier to read than `&&`, also it would be easier to investigate assertion failure if each assert is about one single postcondition